File: src/cppcache/src/statistics/StatisticsTypeImpl.cpp
/*=========================================================================
* Copyright (c) 2004-2014 Pivotal Software, Inc. All Rights Reserved.
 * This product is protected by U.S. and international copyright
 * and intellectual property laws. Pivotal products are covered by
 * more patents listed at http://www.pivotal.io/patents.
*=========================================================================
*/

#include "StatisticsTypeImpl.hpp"
#include "StatisticDescriptorImpl.hpp"
#include <string>
#include <ace/OS.h>
using namespace gemfire_statistics;

/**
* Gathers together a number of {@link StatisticDescriptor statistics}
* into one logical type.
*
*/

/**
* Creates a new <code>StatisticsType</code> with the given name,
* description, and statistics.
*
* @param name
*        The name of this statistics type (for example,
*        <code>"DatabaseStatistics"</code>)
* @param description
*        A description of this statistics type (for example,
*        "Information about the application's use of the
*        database").
* @param stats
*        Descriptions of the individual statistics grouped together
*        in this statistics type{@link StatisticDescriptor}.
*
* @throws NullPointerException
*         If either <code>name</code> or <code>stats</code> is
*         <code>null</code>.
*/

StatisticsTypeImpl::StatisticsTypeImpl(const char* nameArg,
                                       const char* descriptionArg,
                                       StatisticDescriptor** statsArg,
                                       int32 statsLengthArg) {
  if (strcmp(nameArg, "") == 0) {
    const char* s = "Cannot have a null statistics type name";
    throw NullPointerException(s);
  }
  if (statsArg == NULL) {
    const char* s = "Cannot have a null statistic descriptors";
    throw NullPointerException(s);
  }
  if (statsLengthArg > MAX_DESCRIPTORS_PER_TYPE) {
    char buffer[100];
    ACE_OS::snprintf(buffer, 100, "%d", statsLengthArg);
    std::string temp(buffer);
    std::string s = "The requested descriptor count " + temp +
                    " exceeds the maximum which is ";
    ACE_OS::snprintf(buffer, 100, "%d", MAX_DESCRIPTORS_PER_TYPE);
    std::string buf(buffer);
    s += buf + ".";
    throw IllegalArgumentException(s.c_str());
  }
  this->name = nameArg;
  this->description = descriptionArg;
  this->stats = statsArg;
  this->statsLength = statsLengthArg;
  int32 intCount = 0;
  int32 longCount = 0;
  int32 doubleCount = 0;
  for (int32 i = 0; i < this->statsLength; i++) {
    // Concrete class required to set the ids only.
    StatisticDescriptorImpl* sd =
        dynamic_cast<StatisticDescriptorImpl*>(stats[i]);
    /* adongre
     * CID 28694: Dereference after null check (FORWARD_NULL)Comparing "sd" to
     * null implies that "sd" might be null.
     *
     * FIX : Do not check null from the dynamic_cast
     *       catch the exception thrown by the dynamic_cast
     */

    if (sd != NULL) {
      if (sd->getTypeCode() == INT_TYPE) {
        sd->setId(intCount);
        intCount++;
      } else if (sd->getTypeCode() == LONG_TYPE) {
        sd->setId(longCount);
        longCount++;
      } else if (sd->getTypeCode() == DOUBLE_TYPE) {
        sd->setId(doubleCount);
        doubleCount++;
      }
      std::string str = stats[i]->getName();
      StatisticsDescMap::iterator iterFind = statsDescMap.find(str);
      if (iterFind != statsDescMap.end()) {
        std::string s = "Duplicate StatisticDescriptor named ";
        s += sd->getName();
        throw IllegalArgumentException(s.c_str());
      } else {
        // statsDescMap.insert(make_pair(stats[i]->getName(), stats[i]));
        statsDescMap.insert(
            StatisticsDescMap::value_type(stats[i]->getName(), stats[i]));
      }
    }
  }  // for
  this->intStatCount = intCount;
  this->longStatCount = longCount;
  this->doubleStatCount = doubleCount;
}

///////////////////////////////Dtor/////////////////////////

StatisticsTypeImpl::~StatisticsTypeImpl() {
  try {
    // Delete the descriptor pointers from the array
    for (int32 i = 0; i < statsLength; i++) {
      delete stats[i];
      stats[i] = NULL;
    }
    // same pointers are also stored in this map.
    // So, Set the pointers to null.
    StatisticsDescMap::iterator iterFind = statsDescMap.begin();
    while (iterFind != statsDescMap.end()) {
      (*iterFind).second = NULL;
      iterFind++;
    }
  } catch (...) {
  }
}

//////////////////////  StatisticsType Methods  //////////////////////

const char* StatisticsTypeImpl::getName() { return name.c_str(); }

const char* StatisticsTypeImpl::getDescription() { return description.c_str(); }

StatisticDescriptor** StatisticsTypeImpl::getStatistics() { return stats; }

int32 StatisticsTypeImpl::nameToId(const char* nameArg) {
  return nameToDescriptor(nameArg)->getId();
}

StatisticDescriptor* StatisticsTypeImpl::nameToDescriptor(const char* nameArg) {
  std::string str(nameArg);
  StatisticsDescMap::iterator iterFind = statsDescMap.find(str);
  if (iterFind == statsDescMap.end()) {
    std::string statName(nameArg);
    std::string s = "There is no statistic named " + statName +
                    " in this statistics instance ";
    LOGWARN("StatisticsTypeImpl::nameToDescriptor %s", s.c_str());
    throw IllegalArgumentException(s.c_str());
  } else {
    return (*iterFind).second;
  }
}
//////////////////////  Instance Methods  //////////////////////

/**
* Gets the number of statistics in this type that are ints.
*/
int32 StatisticsTypeImpl::getIntStatCount() { return intStatCount; }

/**
* Gets the number of statistics in this type that are longs.
*/
int32 StatisticsTypeImpl::getLongStatCount() { return longStatCount; }

/**
* Gets the number of statistics that are doubles.
*/
int32 StatisticsTypeImpl::getDoubleStatCount() { return doubleStatCount; }

/**
* Gets the total number of statistic descriptors.
*/
int32 StatisticsTypeImpl::getDescriptorsCount() { return statsLength; }

File: geode-core/src/main/java/org/apache/geode/internal/statistics/StatisticsTypeImpl.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
 * agreements. See the NOTICE file distributed with this work for additional information regarding
 * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance with the License. You may obtain a
 * copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package org.apache.geode.internal.statistics;

import java.io.Reader;
import java.util.HashMap;

import org.apache.geode.StatisticDescriptor;
import org.apache.geode.Statistics;
import org.apache.geode.StatisticsType;
import org.apache.geode.StatisticsTypeFactory;
import org.apache.geode.annotations.Immutable;

/**
 * Gathers together a number of {@link StatisticDescriptor statistics} into one logical type.
 *
 * @see Statistics
 *
 *
 * @since GemFire 3.0
 */
@Immutable
public class StatisticsTypeImpl implements ValidatingStatisticsType {

  /** The name of this statistics type */
  private final String name;

  /** The description of this statistics type */
  private final String description;

  /** The descriptions of the statistics in id order */
  private final StatisticDescriptor[] stats;

  /** Maps a stat name to its StatisticDescriptor */
  private final HashMap<String, StatisticDescriptor> statsMap;

  /** Contains the number of long statistics in this type. */
  private final int longStatCount;

  /** Contains the number of double statistics in this type. */
  private final int doubleStatCount;

  ///////////////////// Static Methods /////////////////////

  /**
   * @see StatisticsTypeXml#read(Reader, StatisticsTypeFactory)
   */
  public static StatisticsType[] fromXml(Reader reader, StatisticsTypeFactory factory) {
    return (new StatisticsTypeXml()).read(reader, factory);
  }

  ////////////////////// Constructors //////////////////////

  /**
   * Creates a new <code>StatisticsType</code> with the given name, description, and statistics.
   *
   * @param name The name of this statistics type (for example, <code>"DatabaseStatistics"</code>)
   * @param description A description of this statistics type (for example, "Information about the
   *        application's use of the database").
   * @param stats Descriptions of the individual statistics grouped together in this statistics
   *        type.
   *
   * @throws NullPointerException If either <code>name</code> or <code>stats</code> is
   *         <code>null</code>.
   */
  public StatisticsTypeImpl(String name, String description, StatisticDescriptor[] stats) {
    if (name == null) {
      throw new NullPointerException(
          "Cannot have a null statistics type name.");
    }

    if (stats == null) {
      throw new NullPointerException(
          "Cannot have a null statistic descriptors.");
    }
    if (stats.length > StatisticsTypeFactory.MAX_DESCRIPTORS_PER_TYPE) {
      throw new IllegalArgumentException(
          String.format("The requested descriptor count %s exceeds the maximum which is  %s .",
              stats.length,
              StatisticsTypeFactory.MAX_DESCRIPTORS_PER_TYPE));
    }

    this.name = name;
    this.description = description;
    this.stats = stats;
    statsMap = new HashMap<>(stats.length * 2);

    longStatCount = addTypedDescriptorToMap(StatisticDescriptorImpl.LONG, 0);
    doubleStatCount = addTypedDescriptorToMap(StatisticDescriptorImpl.DOUBLE, longStatCount);
  }

  private int addTypedDescriptorToMap(byte typeCode, int startOfSequentialIds) {
    int count = 0;
    for (StatisticDescriptor stat : stats) {
      StatisticDescriptorImpl sd = (StatisticDescriptorImpl) stat;
      if (sd.getTypeCode() == typeCode) {
        sd.setId(startOfSequentialIds + count);
        count++;
        addDescriptorToMap(sd);
      }
    }
    return count;
  }

  private void addDescriptorToMap(StatisticDescriptor sd) {
    Object previousValue = statsMap.put(sd.getName(), sd);
    if (previousValue != null) {
      throw new IllegalArgumentException(
          String.format("Duplicate StatisticDescriptor named %s",
              sd.getName()));
    }
  }

  ////////////////////// StatisticsType Methods //////////////////////

  @Override
  public String getName() {
    return name;
  }

  @Override
  public String getDescription() {
    return description;
  }

  @Override
  public StatisticDescriptor[] getStatistics() {
    return stats;
  }

  @Override
  public int nameToId(String name) {
    return nameToDescriptor(name).getId();
  }

  @Override
  public StatisticDescriptor nameToDescriptor(String name) {
    StatisticDescriptor stat = statsMap.get(name);
    if (stat == null) {
      throw new IllegalArgumentException(
          String.format("There is no statistic named %s",
              name));
    }
    return stat;
  }

  ////////////////////// Instance Methods //////////////////////

  /**
   * Gets the number of statistics in this type that are longs.
   */
  public int getLongStatCount() {
    return longStatCount;
  }

  /**
   * Gets the number of statistics that are doubles.
   */
  public int getDoubleStatCount() {
    return doubleStatCount;
  }

  // @Override
  // public String toString() {
  // return "StatisticType with " + this.stats.length + " stats";
  // }

  @Override
  public String toString() {
    final StringBuilder sb = new StringBuilder(getClass().getName());
    sb.append("@").append(System.identityHashCode(this)).append("{");
    sb.append("name=").append(name);
    sb.append(", description=").append(description);
    sb.append(", stats.length=").append(stats.length);
    sb.append("}");
    return sb.toString();
  }

  @Override
  public int hashCode() {
    return getName().hashCode();
  }

  @Override
  public boolean equals(Object o) {
    if (o == null) {
      return false;
    }
    if (!(o instanceof StatisticsType)) {
      return false;
    }
    StatisticsType other = (StatisticsType) o;
    if (!getName().equals(other.getName())) {
      return false;
    }
    if (!getDescription().equals(other.getDescription())) {
      return false;
    }
    StatisticDescriptor[] myStats = getStatistics();
    StatisticDescriptor[] yourStats = other.getStatistics();
    if (myStats.length != yourStats.length) {
      return false;
    }
    for (int i = 0; i < myStats.length; i++) {
      if (!myStats[i].equals(yourStats[i])) {
        return false;
      }
    }
    return true;
  }

  @Override
  public boolean isValidLongId(int id) {
    return id < longStatCount;
  }

  @Override
  public boolean isValidDoubleId(int id) {
    return longStatCount <= id && id < longStatCount + doubleStatCount;
  }
}
